add anylabeling interactive tool

sunyi000 opened the pull request:
Adding Anylabeling interactive tool.
https://github.com/vietanhdev/anylabeling

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

Commit:
anylabeling

Commit:
add classes

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,71 @@
+<tool id="interactive_tool_anylabeling" tool_type="interactive" name="AnyLabeling Interactive" version="@VERSION@" profile="24.2">
+    <description>Effortless data labeling with AI support from YOLO and Segment Anything! AnyLabeling = LabelImg + Labelme + Improved UI + Auto-labeling</description>
```
```suggestion
    <description>Effortless data labeling with AI support from YOLO and Segment Anything</description>
```

maybe put that in the help?

bgruening replied:
Do you mind adding an icon to the IT (and others if you like).

```suggestion
    <description>Effortless data labeling with AI support from YOLO and Segment Anything! AnyLabeling = LabelImg + Labelme + Improved UI + Auto-labeling</description>
    <icon src="anylabeling.png" />
```

For example, the icon needs to be next to the XML

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,71 @@
+<tool id="interactive_tool_anylabeling" tool_type="interactive" name="AnyLabeling Interactive" version="@VERSION@" profile="24.2">
+    <description>Effortless data labeling with AI support from YOLO and Segment Anything! AnyLabeling = LabelImg + Labelme + Improved UI + Auto-labeling</description>
+    <requirements>
+	    <container type="docker">quay.io/sunyi000/docker-yololabel:v0.4.16</container>
```
Do you want to do the same than last time? The github repo.

sunyi000 replied:
yes I can do the same as last time.

bgruening replied:
https://github.com/usegalaxy-eu/docker-anylabeling here it is.

sunyi000 replied:
Thanks @bgruening ,I pushed the files, added the .github/workflows ymls and made a release, but the repo in quay.io still does not exist..is there anything else i need to do to make it appear? should i make the release.yml point to my own quay.io repo?

bgruening replied:
I think the CI yml files needs to be in a .github/workflows/ folder

sunyi000 replied:
Hi @bgruening am I still doing something wrong, the CI job now runs successfully, but the repo is not created in quay.io. Thank you

bgruening replied:
https://quay.io/repository/bgruening/anylabeling?tab=settings

Quay.io makes reos hidden by default. I changed it to public. Thanks @sunyi000

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,71 @@
+	 chown -R 1000:1000 ./home &&
+
+         #for $i,$img in enumerate($input_images):
+            ln -s "$img" "./home/input_images/${img.element_identifier}" &&
```
Please use single-quotes.

```suggestion
            ln -s '$img' './home/input_images/${img.element_identifier}' &&
```

nilchia commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,71 @@
+        <param name="input_images" type="data" format="jpg,png,tiff,bmp,gif,pcx,ppm,psd,pbm,pgm,eps" multiple="true"  label="Input images"/>
+        <param name="annotations" type="data" format="txt,json" optional="true" multiple="true" label="Pre-existing annotations"/>
+        <param name="classes" type="data" format="txt" optional="true" label="Classes file (one class per line)"/>
+        <param name="models" type="data" format="auto" optional="true" label="Auto-labeling model file"/>
```
```suggestion
        <param name="classes" type="data" format="txt" optional="true" multiple="true" label="Classes file (one class per line)"/>
        <param name="models" type="data" format="auto" optional="true" multiple="true" label="Auto-labeling model file"/>
```

nilchia replied:
I think those also need to be `multiple="true"` as you are using `enumerate`.

sunyi000 replied:
thank you. just updated them.

sunyi000 committed:
Update tools/interactive/interactivetool_anylabeling.xml

Co-authored-by: Amirhossein Nilchi <66441226+nilchia@users.noreply.github.com>

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,71 @@
+        <param name="input_images" type="data" format="jpg,png,tiff,bmp,gif,pcx,ppm,psd,pbm,pgm,eps" multiple="true"  label="Input images"/>
+        <param name="annotations" type="data" format="txt,json" optional="true" multiple="true" label="Pre-existing annotations"/>
+        <param name="classes" type="data" format="txt" optional="true" label="Classes file (one class per line)"/>
+        <param name="models" type="data" format="auto" optional="true" label="Auto-labeling model file"/>
```
what are those models rechnically?

sunyi000 replied:
it will be a mix of onnx model files and yml files, as described here https://anylabeling.nrl.ai/docs/custom-models

bgruening replied:
In theory there is a dicrectory datatype in Galaxy. But I guess its easier if users provide the models as tar?

sunyi000 replied:
i have changed this to tar format, so it takes 1 tar of model files and yaml files..

Commit:
updates, add icon etc

Commit:
change models to tar ball

Commit:
change docker repo on quay.io

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,83 @@
+    <![CDATA[
+         export HOME=\$PWD &&
+         
+         mkdir -p ./home ./home/output ./home/input_images &&
```
```suggestion
         mkdir -p ./home/output ./home/input_images &&
```

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,83 @@
+         export HOME=\$PWD &&
+         
+         mkdir -p ./home ./home/output ./home/input_images &&
+	 chown -R 1000:1000 ./home &&
```
why 1000?

sunyi000 replied:
It is from the docker base image. It's the default ID of the user the application runs as. I saw the ilastik and other IT also have this

bgruening replied:
Just wanted to make sure it's needed.

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,83 @@
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff,bmp,gif,pcx,ppm,psd,pbm,pgm,eps" multiple="true"  label="Input images"/>
+        <param name="annotations" type="data" format="txt,json" optional="true" multiple="true" label="Pre-existing annotations"/>
+        <param name="classes" type="data" format="txt" optional="true" multiple="true" label="Classes file (one class per line)"/>
```
Is that just one file or multiple files. The lable indicates its one file.

Is that maybe one classfile per input image?

Or maybe a more general Q can I have X input images, Y annotations and Z classes?

sunyi000 replied:
One image 1 annotation file, but multiple images share the same class file. may be the class param should be single...

bgruening commented on tools/interactive/interactivetool_anylabeling.xml:
```diff
@@ -0,0 +1,83 @@
+        <param name="input_images" type="data" format="jpg,png,tiff,bmp,gif,pcx,ppm,psd,pbm,pgm,eps" multiple="true"  label="Input images"/>
+        <param name="annotations" type="data" format="txt,json" optional="true" multiple="true" label="Pre-existing annotations"/>
+        <param name="classes" type="data" format="txt" optional="true" multiple="true" label="Classes file (one class per line)"/>
+        <param name="models" type="data" format="tar" optional="true" label="A tarball containing custom model files and yaml files"/>
```
It would be helpful to mention maybe in the tool help where you get the models.

sunyi000 replied:
added, together with changing the class input to single. could you do another round of review, @bgruening Thank you

sunyi000 committed:
Update tools/interactive/interactivetool_anylabeling.xml

Co-authored-by: Björn Grüning <bjoern@gruenings.eu>

Commit:
add help, change class file param to single input

bgruening commented:
Thanks. Will you test it on Monday?

sunyi000 commented:
> Thanks. Will you test it on Monday?

yes I will test it on Monday and let you know.
Have a good weekend!

bgruening commented:
You too!

sunyi000 commented:
> You too!

Hi @bgruening , I juste tested it in usegalaxy.eu

Locally, If I did the below, it works: 
`docker pull quay.io/bgruening/anylabeling:0.4.30 `
and then
`docker run -e DISPLAY=$DISPLAY -p 5800:5800 89ab161b2dcc`

On usegalaxy.eu
The tool is running, but, if I click the button to open it in a separate tab, I get
- Proxy target missing
- in the log
```
[supervisor  ] starting service 'openbox'...
[openbox     ] Fontconfig warning: ignoring en      NUMBA_CACHE_DIR /tmp: not a valid language tag
[supervisor  ] starting service 'xcompmgr'...
[supervisor  ] starting service 'app'...
[app         ] Failed to save config: /data/jwd06/pulsar_staging/83395010/working/.anylabelingrc
[supervisor  ] all services started.
[app         ] Traceback (most recent call last):
[app         ]   File "/opt/conda/envs/anylabeling/bin/anylabeling", line 8, in <module>
[app         ]     sys.exit(main())
[app         ]              ^^^^^^
[app         ]   File "/opt/conda/envs/anylabeling/lib/python3.12/site-packages/anylabeling/app.py", line 154, in main
[app         ]     config = get_config(config_file_or_yaml, config_from_args)
[app         ]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[app         ]   File "/opt/conda/envs/anylabeling/lib/python3.12/site-packages/anylabeling/config.py", line 80, in get_config
[app         ]     with open(config_from_yaml) as f:
[app         ]          ^^^^^^^^^^^^^^^^^^^^^^
[app         ] FileNotFoundError: [Errno 2] No such file or directory: '/data/jwd06/pulsar_staging/83395010/working/.anylabelingrc'
```

bgruening commented:
Looks like some permission issues to me. 

Try to run it locally with ` docker run -e DISPLAY=$DISPLAY -p 5800:5800 --user 1001:1001 89ab161b2dcc` or any random user id, can you then reproduce it?

sunyi000 commented:
> Looks like some permission issues to me.
> 
> Try to run it locally with ` docker run -e DISPLAY=$DISPLAY -p 5800:5800 --user 1001:1001 89ab161b2dcc` or any random user id, can you then reproduce it?

Hi @bgruening , I looked a bit more, and it seems that the tool is trying to create a folder upon starting, but it does not have permission, because it starts as the galaxy logged in user, but the user does not have permission write on the galaxy server?
From the anylabeling side, it seems the directory is not configurable, so I can not create something else in the xml wrapper and point anylabeling to it. if this is the case what can I do to workaround this?
see below stack trace

```
    self.parent.mkdir(parents=True, exist_ok=True)
[app         ]   File "/opt/conda/envs/anylabeling/lib/python3.12/pathlib.py", line 1311, in mkdir
[app         ]     os.mkdir(self, mode)
[app         ] PermissionError: [Errno 13] Permission denied: '/data/jwd06/pulsar_staging/83420322/working/anylabeling_data'
```

thank you very much

bgruening commented:
Do you know where it writes to? Is that $HOME? or the folder where the application is started? Can you set $HOME ?

sunyi000 commented:
> Do you know where it writes to? Is that $HOME? or the folder where the application is started? Can you set $HOME ?

From what I can find, it creates the folder in `current working directory`. The same place where I did `mkdir ./home/input` in the xml.
I already set $HOME=\$PWD

sunyi000 commented:
Hi @bgruening , 
I somehow made it working now in usegalaxy.eu
I modified the Dockerfile, changed 1 line of the AnyLabeling's source code. although am not 100% sure if this change made it work :)
https://github.com/usegalaxy-eu/docker-anylabeling/blob/55bd2dc9d3ed3aa57ee1171ac38b11743b494db5/Dockerfile#L66
It is probably not ideal and a bit hacky, but works for now. 
I also opened a issue in AnyLabeling git repo, asking the possibility to make the folder location configurable.

bgruening commented:
Hacker! :)

Can you maybe add a comment to the Dockerfile, why this is needed and link to the upstream issue?

So its working now? That is super cool! Congratulastions! If you think its already useful for users, please consider writing a small news item with a screenshot - then NFDI is also happy :)

sunyi000 commented:
> Hacker! :)
> 
> Can you maybe add a comment to the Dockerfile, why this is needed and link to the upstream issue?
> 
> So its working now? That is super cool! Congratulastions! If you think its already useful for users, please consider writing a small news item with a screenshot - then NFDI is also happy :)

I added comments to Dockerfile.
Yes, I tested it with a small example, and it worked.
i will write a small blog for it similar to what i did for the napari
thank you!

bgruening commented:
Thanks to you!
